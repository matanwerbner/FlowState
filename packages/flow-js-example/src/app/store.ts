import { create } from 'zustand';
import {
  InitialData,
  InitialState,
  createStateMachine,
} from './lib/createStateMachine';
import { EventNames, GameContext, GameStates } from './lib/types';

type InitializeParams<GameContext> = {
  state?: GameStates;
  data?: Partial<GameContext>;
};

type Storetype = {
  currentState: GameStates;
  triggerEvent: (eventName: EventNames, data?: Partial<GameContext>) => void;
  context: GameContext;
  initializeStore: ({ state, data }: InitializeParams<GameContext>) => void;
};

export const useGameStore = create<Storetype>((set) => {
  const stateMachine = createStateMachine();

  return {
    context: stateMachine.context,
    currentState: stateMachine.currentState,
    triggerEvent: (eventName: EventNames, data?: Partial<GameContext>) => {
      return stateMachine.triggerEvent(eventName, data).then((newState) => {
        set({ currentState: newState, context: stateMachine.context });
      });
    },
    initializeStore: ({ state = InitialState, data = InitialData }) => {
      stateMachine.initialize(state, { ...InitialData, ...data });
      set({
        currentState: stateMachine.currentState,
        context: stateMachine.context,
      });
    },
  };
});
