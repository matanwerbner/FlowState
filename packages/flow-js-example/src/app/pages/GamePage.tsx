import styled from 'styled-components';
import { FlexBox } from '../components/FlexBox';
import { useGameStore } from '../store';
import { sanitize } from 'dompurify';
import { useState } from 'react';
import { Button } from '../components/Button';
import { AnswesContainer } from './GampePage.styled';

const AnimtionDurationMS = 500;

export const GamePage = () => {
  const context = useGameStore((state) => state.context);
  const question = context.questions[context.currentQuestionIndex];
  const triggerEvent = useGameStore((state) => state.triggerEvent);
  const [isTransitioning, setIsTransitioning] = useState(false);

  const handleAnswer = (answer: boolean) => {
    setIsTransitioning(true);
    setTimeout(() => {
      triggerEvent('answerQuestion', {
        ...(question.answer === answer
          ? { correctAnswers: context.correctAnswers + 1 }
          : { incorrectAnswers: context.incorrectAnswers + 1 }),
      });
      if (context.currentQuestionIndex + 1 < context.questions.length) {
        setIsTransitioning(false);
      }
    }, AnimtionDurationMS);
  };

  return (
    <FlexBox style={{ opacity: isTransitioning ? 0 : 1 }}>
      <h3 dangerouslySetInnerHTML={{ __html: sanitize(question.question) }} />
      <AnswesContainer>
        <Button color="primary" onClick={() => handleAnswer(true)}>
          True
        </Button>
        <Button color="secondary" onClick={() => handleAnswer(false)}>
          False
        </Button>
      </AnswesContainer>
    </FlexBox>
  );
};
