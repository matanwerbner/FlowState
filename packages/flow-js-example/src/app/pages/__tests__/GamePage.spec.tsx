import { render } from '@testing-library/react';
import { GamePage } from '../GamePage';
import { useGameStore } from '../../store';

describe('GamePage tests', () => {
  it('should render', () => {
    useGameStore.setState({
      context: {
        questions: [{ question: 'test', answer: true }],
        currentQuestionIndex: 0,
        correctAnswers: 0,
        incorrectAnswers: 0,
        category: 'test',
        difficulty: 'easy',
      },
    });
    const { baseElement } = render(<GamePage />);
    expect(baseElement).toMatchSnapshot();
  });
});
