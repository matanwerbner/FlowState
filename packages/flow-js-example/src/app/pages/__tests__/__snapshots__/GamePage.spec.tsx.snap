// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GamePage tests should render 1`] = `
<body>
  <div>
    <div
      class="FlexBox__StyledBox-txfyk3-0 lgdOeB MuiBox-root css-329zn9"
      style="opacity: 1;"
    >
      <h3>
        test
      </h3>
      <div
        class="GampePagestyled__AnswesContainer-sc-4idwn3-0 bMINnz"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorPrimary css-sghohy-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          True
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSecondary MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-colorSecondary css-zcbmsk-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          False
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
  </div>
</body>
`;
